internal/helm/repository/chart_repository.go

```diff
@@ -288,155 +298,86 @@ func (r *ChartRepository) DownloadChart(chart *repo.ChartVersion) (*bytes.Buffer
-	}
-	b, err := os.ReadFile(path)
-	if err != nil {
-		return err
-	}
-	return r.LoadIndexFromBytes(b)
-}
-
 // CacheIndex attempts to write the index from the remote into a new temporary file
-// using DownloadIndex, and sets CachePath and Cached.
+// using DownloadIndex, and sets Path and cached.
 // It returns the SHA256 checksum of the downloaded index bytes, or an error.
-// The caller is expected to handle the garbage collection of CachePath, and to
-// load the Index separately using LoadFromCache if required.
-func (r *ChartRepository) CacheIndex() (string, error) {
+// The caller is expected to handle the garbage collection of Path, and to
+// load the Index separately using LoadFromPath if required.
+func (r *ChartRepository) CacheIndex() error {
 	f, err := os.CreateTemp("", "chart-index-*.yaml")
 	if err != nil {
-		return "", fmt.Errorf("failed to create temp file to cache index to: %w", err)
+		return fmt.Errorf("failed to create temp file to cache index to: %w", err)
 	}
 
-	h := sha256.New()
-	mw := io.MultiWriter(f, h)
-	if err = r.DownloadIndex(mw); err != nil {
+	if err = r.DownloadIndex(f); err != nil {
 		f.Close()
-		os.RemoveAll(f.Name())
-		return "", fmt.Errorf("failed to cache index to temporary file: %w", err)
+		os.Remove(f.Name())
+		return fmt.Errorf("failed to cache index to temporary file: %w", err)
 	}
 	if err = f.Close(); err != nil {
-		os.RemoveAll(f.Name())
-		return "", fmt.Errorf("failed to close cached index file '%s': %w", f.Name(), err)
+		os.Remove(f.Name())
+		return fmt.Errorf("failed to close cached index file '%s': %w", f.Name(), err)
 	}
 
 	r.Lock()
-	r.CachePath = f.Name()
-	r.Cached = true
+	r.Path = f.Name()
+	r.Index = nil
+	r.cached = true
+	r.invalidate()
 	r.Unlock()
-	return hex.EncodeToString(h.Sum(nil)), nil
-}
-
-// CacheIndexInMemory attempts to cache the index in memory.
-// It returns an error if it fails.
-// The cache key have to be safe in multi-tenancy environments,
-// as otherwise it could be used as a vector to bypass the helm repository's authentication.
-func (r *ChartRepository) CacheIndexInMemory() error {
-	// Cache the index if it was successfully retrieved
-	// and the chart was successfully built
-	if r.IndexCache != nil && r.Index != nil {
-		err := r.IndexCache.Set(r.IndexKey, r.Index, r.IndexTTL)
-		if err != nil {
-			return err
-		}
-	}
 
 	return nil
 }
 
-// StrategicallyLoadIndex lazy-loads the Index
-// first from Indexcache,
-// then from CachePath using oadFromCache if it does not HasIndex.
-// If not HasCacheFile, a cache attempt is made using CacheIndex
-// before continuing to load.
+// StrategicallyLoadIndex lazy-loads the Index if required, first
+// attempting to load it from Path if the file exists, before falling
+// back to caching it.
 func (r *ChartRepository) StrategicallyLoadIndex() (err error) {
 	if r.HasIndex() {
 		return
 	}
 
-	if r.IndexCache != nil {
```

Comment:
This is one of the central point where the `in-memory cache` is used.

Building chart from `remote` does the following:
1. `remote.GetChartVersion` https://github.com/fluxcd/source-controller/blob/5a01112253cc41478163aa5c487afa77c61851d1/internal/helm/chart/builder_remote.go#L129
2. `remote.DownloadChart()`

In order to get the chart version to download, `remote.GetChartVersion` calls `StrategicallyLoadIndex()`which first attempt to get it from the `in-memory cache`, then from `path` and finally download it.